test/test_infer.rb

```diff
@@ -8,6 +8,60 @@
 require 'rdl'
 require 'types/core'
 
+class CompareTypeVariables
+  def initialize
+    @var_lookup = {}
+  end
+
+  def compare_methods(expected, given)
+    expected.args.length == given.args.length &&
+      expected.args.zip(given.args).all? { |a| compare(a[0], a[1]) } &&
+      compare(expected.block, given.block) &&
+      compare(expected.ret, given.ret)
+  end
+
+  def compare_structural(expected, given)
+    expected.methods.keys == given.methods.keys &&
+      expected.methods.values.zip(given.methods.values).all? { |a| compare(a[0], a[1]) }
+  end
+
+  def compare_hash(expected, given)
+    expected.elts.keys == given.elts.keys &&
+      expected.elts.values.zip(given.elts.values).all? { |a| compare(a[0], a[1]) }
+  end
+
+  def compare_vartype(expected, given)
+    name_sym = expected.name.to_sym
+    return @var_lookup[name_sym] == given if @var_lookup.key? name_sym
```

Comment:
So the basic way this works is to walk the type-trees of both the expected and inferred types together, comparing them as they go. When it encounters a type variable in `expected` that it hasn't seen yet, it stores the `given` type variable and returns `true`. If it has seen it, it compares what it has in its table and returns the result of that.

In the test `plus_two`, the output of the function looks like a different type than the output of `var#+`, for reasons of how inference reports the solutions of type variables---when in fact, b ≤ a. The expected type right now is `([ +: (Number) -> a ]) -> b`. It seems like too permissive of a test to say that these two things are completely unrelated, so one thing we could do here to test this is change `@var_lookup[name_sym] == given` to `@var_lookup[name_sym] <= given`. This would make the expected type for the `plus_two` test `([ +: (Number) -> a ]) -> a`. I'm not sure how else we'd represent that in these tests without changing the type grammar.